core/src/sumeragi/mod.rs

```diff
@@ -246,9 +246,38 @@ impl Sumeragi {
     pub fn initialize_and_start_thread(
         sumeragi: Arc<Self>,
         genesis_network: Option<GenesisNetwork>,
+        block_hashes: &[HashOf<VersionedCommittedBlock>],
     ) -> ThreadHandler {
         let wsv = sumeragi.wsv_mutex_access().clone();
 
+        for (block_hash, i) in block_hashes
+            .iter()
+            .take(block_hashes.len().saturating_sub(1))
+            .zip(1u64..)
+        {
+            let block_height: u64 = i;
+            let block_ref = sumeragi.internal.kura.get_block_by_height(block_height).expect("Sumeragi could not load block that was reported as present. Please check that the block storage was not disconnected.");
+            assert_eq!(
+                block_ref.hash(),
+                *block_hash,
+                "Kura init correctly reported the block hash."
+            );
+
+            wsv.apply(&block_ref)
+                .expect("Failed to apply block to wsv in init.");
+        }
+        let finalized_wsv = wsv.clone();
+
+        if !block_hashes.is_empty() {
+            let block_ref = sumeragi.internal.kura.get_block_by_height(block_hashes.len() as u64).expect("Sumeragi could not load block that was reported as present. Please check that the block storage was not disconnected.");
```

Comment:
shouldn't you be getting the last block from `kura`? If you're getting a block by it's height then it's not guaranteed you will get the last block

Reply:
The block hashes passed into this function was just produced by kura and contains all the blocks in the store.